Feature/3 drtma3

EdwardColon-NOAA opened the pull request:
The remaining novel SPC fields are being submitted for review. These include: Effective bulk shear, U and V components of effective shear, U and V Components of effective Bunkers storm motion, effective helicity, effective sig tornado parameter, fixed layer sig tornado parameter, supercell parameter, hail parameter, and mixed layer virtual LFC.

EdwardColon-NOAA committed:
Adding Effective Layer fields including: EL top and bottom, Temperature at EL top, U and V components of EL shear, and EL shear wind magnitude. Changes are confined to subprograms RQSTFLD.F and MISCLN.f. Passed regression tests.

EdwardColon-NOAA committed:
Merge branch 'develop' into feature/3DRTMA2

EdwardColon-NOAA committed:
Update MISCLN.f

EdwardColon-NOAA committed:
Update post_avblflds.xml

EdwardColon-NOAA committed:
Delete post_avblflds_3drtma.xml

EdwardColon-NOAA committed:
Update makefile

EdwardColon-NOAA committed:
Update params_grib2_tbl_new.text

EdwardColon-NOAA committed:
Update params_grib2_tbl_new

EdwardColon-NOAA committed:
Update 3drtma_postcntrl.xml

EdwardColon-NOAA committed:
Update 3drtma_postcntrl.xml

EdwardColon-NOAA committed:
Create postxconfig-NT-3drtma.txt

EdwardColon-NOAA committed:
fixed the truncation of parm/postxconfig-NT-3drtma.txt.

EdwardColon-NOAA committed:
Updating UPP configuration files for 3D-RTMA case.

EdwardColon-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/3DRTMA2

EdwardColon-NOAA committed:
First commit of the remaining SPC variables.

EdwardColon-NOAA committed:
merged key changes back into code.

EdwardColon-NOAA committed:
The remaining novel SPC fields are being submitted for review. These include: Effective bulk shear, U and V components of effective shear, U and V Components of effective Bunkers storm motion, effective helicity, effective sig tornado parameter, fixed layer sig tornado parameter, supercell parameter, hail parameter, and mixed layer virtual LFC.

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -282,7 +295,13 @@ SUBROUTINE MISCLN
          CALL CALHEL(DEPTH,UST,VST,HELI,USHR1,VSHR1,USHR6,VSHR6)
-
+! 0-6 km shear magnitude
+!$omp parallel do private(i,j,jj)
```
Remove "jj" for private.

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3127,27 +3152,27 @@ SUBROUTINE MISCLN
+                 GRID1(I,J)=EGRID2(I,J)
+                 IF (SUBMODELNAME == 'RTMA') MLLCL(I,J) = GRID1(I,J)
+               ENDDO
+               ENDDO
```
Could you document why this block is uncommented out in giithub issue?

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3618,46 +3674,6 @@ SUBROUTINE MISCLN
              endif
             ENDIF
-!Temperature of effbot
-            IF (IGET(981)>0) THEN
```
Please document why ID 981 is removed in github issue.

EdwardColon-NOAA committed:
Update MISCLN.f

EdwardColon-NOAA committed:
Update post_avblflds.xml

EdwardColon-NOAA committed:
Update 3drtma_postcntrl.xml

EdwardColon-NOAA committed:
Update postxconfig-NT-3drtma.txt

WenMeng-NOAA commented:
@EdwardColon-NOAA Could you sync your branch with the latest changes from upstream/develop? I will test your PR being next merging.

EdwardColon-NOAA commented:
Sorry for the delay. Will do.

-Edward

On Thu, Mar 25, 2021 at 11:16 AM WenMeng-NOAA ***@***.***>
wrote:

> @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA> Could you sync
> your branch with the latest changes from upstream/develop? I will test your
> PR being next merging.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/280#issuecomment-806930002>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQGTU6OT4IC5HFFKENDTFNHS5ANCNFSM4ZHVHYSQ>
> .
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/3DRTMA3

GangZhao-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3736,6 +3752,328 @@ SUBROUTINE MISCLN
+            DO J=JSTA,JEND
+               DO I=1,IM
+!               LAPSE=-((T700(I,J)-T500(I,J))/(D50*(Z700(I,J)-Z500(I,J))))
+                LAPSE=(T700(I,J)-T500(I,J)) - 7.
```
Hi Ed,
I remember Annette asked question about the calculation of lapse. Why using "LAPSE=(T700(I,J)-T500(I,J)) - 7.", not the definition which is the commented line above?

GangZhao-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3736,6 +3752,328 @@ SUBROUTINE MISCLN
+             ITYPE = 2
+! EGRID3 is Virtual LFC
+             CALL CALCAPE2(ITYPE,DPBND,P1D,T1D,Q1D,LB2,            &
+                           EGRID1,EGRID2,EGRID3,EGRID4,EGRID5,     &
```
Hi Ed,

Is this Q1D the mixing ratio of the Most Unstable parcel? calculated when ITYPE=2 only?

GangZhao-NOAA replied:
Hi Ed,
For using Q1D here as the Most Unstable mixing ratio, I am not sure about this. 
But to help the merging be done soon, I just comment and have the code pass the review.
We may take a look at it later.
thanks!
- Gang

EdwardColon-NOAA commented:
I'll resolve the problem now, Gang.

-Edward

On Thu, Mar 25, 2021 at 12:40 PM Gang ***@***.***> wrote:

> *@GangZhao-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/MISCLN.f
> <https://github.com/NOAA-EMC/EMC_post/pull/280#discussion_r601658338>:
>
> > +!$omp parallel do private(i,j,jj)
> +               do j=1,jend-jsta+1
> +                 jj = jsta+j-1
> +                 do i=1,im
> +                   datapd(i,j,cfld) = GRID1(i,jj)
> +                 enddo
> +               enddo
> +             endif
> +           ENDIF
> +
> +!Hail parameter
> +            IF (IGET(993)>0) THEN
> +            DO J=JSTA,JEND
> +               DO I=1,IM
> +!               LAPSE=-((T700(I,J)-T500(I,J))/(D50*(Z700(I,J)-Z500(I,J))))
> +                LAPSE=(T700(I,J)-T500(I,J)) - 7.
>
> Hi Ed,
> I remember Annette asked question about the calculation of lapse. Why
> using "LAPSE=(T700(I,J)-T500(I,J)) - 7.", not the definition which is the
> commented line above?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/280#pullrequestreview-621341457>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQDZUTLXHTIZ6UNXRP3TFNRN5ANCNFSM4ZHVHYSQ>
> .
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Update MISCLN.f

GangZhao-NOAA reviewed (approved):
Still have some concern on using Q1D as the mixing ratio of most unstable parcel to compute SHIP in MISCLN.f.
To help the merging be done quickly, I approved this merging here.
We may take a look at this issue later.

Thanks!
- Gang

EdwardColon-NOAA commented:
Hi Gang,

Sorry for the delay.

I'll take a look at the use of Q1D now.

-Edward

On Thu, Mar 25, 2021 at 1:00 PM Gang ***@***.***> wrote:

> *@GangZhao-NOAA* approved this pull request.
>
> Still have some concern on using Q1D as the mixing ratio of most unstable
> parcel to compute SHIP in MISCLN.f.
> To help the merging be done quickly, I approved this merging here.
> We may take a look at this issue later.
>
> Thanks!
>
>    - Gang
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/280#pullrequestreview-621364882>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQBCNCCCTAGDOXVJUB3TFNT3HANCNFSM4ZHVHYSQ>
> .
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA commented:
Hi Gang,

The quantity being sought by SPC is the mixed layer virtual LFC - not most
unstable LFC. So the logic being employed corresponds to that already used
in computing the mixed layer CAPE and CIN. Namely:

!       MIXED LAYER CAPE AND CINS
!
         FIELD1=.FALSE.
         FIELD2=.FALSE.
!
         IF(IGET(032)>0)THEN
           IF(LVLS(3,IGET(032))>0)FIELD1=.TRUE.
         ENDIF
         IF(IGET(107)>0)THEN
           IF(LVLS(3,IGET(107))>0)FIELD2=.TRUE.
         ENDIF
!
         IF(IGET(582)>0)THEN
           FIELD1=.TRUE.
         ENDIF
         IF(IGET(583)>0)THEN
           FIELD2=.TRUE.
         ENDIF

         IF(FIELD1.OR.FIELD2)THEN
           ITYPE = 2
!
!$omp parallel do private(i,j)
           DO J=JSTA,JEND
             DO I=1,IM
               EGRID1(I,J) = -H99999
               EGRID2(I,J) = -H99999
!          ENDDO
!          ENDDO
!          DO J=JSTA,JEND
!          DO I=1,IM




* LB2(I,J)  = (LVLBND(I,J,1) + LVLBND(I,J,2) +           &
          LVLBND(I,J,3))/3               P1D(I,J)  = (PBND(I,J,1) +
PBND(I,J,2) + PBND(I,J,3))/3               T1D(I,J)  = (TBND(I,J,1) +
TBND(I,J,2) + TBND(I,J,3))/3               Q1D(I,J)  = (QBND(I,J,1) +
QBND(I,J,2) + QBND(I,J,3))/3*
             ENDDO
           ENDDO
!

On Thu, Mar 25, 2021 at 1:21 PM Edward Colon - NOAA Affiliate <
***@***.***> wrote:

> Hi Gang,
>
> Sorry for the delay.
>
> I'll take a look at the use of Q1D now.
>
> -Edward
>
> On Thu, Mar 25, 2021 at 1:00 PM Gang ***@***.***> wrote:
>
>> *@GangZhao-NOAA* approved this pull request.
>>
>> Still have some concern on using Q1D as the mixing ratio of most unstable
>> parcel to compute SHIP in MISCLN.f.
>> To help the merging be done quickly, I approved this merging here.
>> We may take a look at this issue later.
>>
>> Thanks!
>>
>>    - Gang
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/EMC_post/pull/280#pullrequestreview-621364882>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACFXXQBCNCCCTAGDOXVJUB3TFNT3HANCNFSM4ZHVHYSQ>
>> .
>>
>
>
> --
> -------------------------------------------------
> Edward Colón
> IMSG at NOAA/NWS/NCEP/EMC
> 5830 University Research Ct.
> Rm #2025
> Riverdale, MD 20737
> Office: (301) 683-3815
> Cell: (301) 213-3566
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

GangZhao-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3736,6 +3752,328 @@ SUBROUTINE MISCLN
+!               LAPSE=-((T700(I,J)-T500(I,J))/(D50*(Z700(I,J)-Z500(I,J))))
+                LAPSE=(T700(I,J)-T500(I,J)) - 7.
+                if (LAPSE <= 0.)LAPSE=0.
+                SHIP=(MUCAPE(I,J)*D1000*Q1D(I,J)*LAPSE*(T500(I,J))*FSHR(I,J))/HCONST
```
Hi Ed,
The Q1D  I mentioned in last comment is this Q1D used to compute SHIP. In last comment, I put the comment to the Q1S used to compute mixed layer (or LFC?)_, sorry for the confusion.
So in this code block to compute SHIP,  where does this Q1D come from? Since no CALCAPE or CALCAPE2 is called in this "IF (IGET(993)>0) THEN" block, so it has to come from somewhere before this IF-block, like just from the IF-block above for IGET(992) for mixed layer (or LFC?).
Thanks!
- Gang

EdwardColon-NOAA commented:
Hi Gang,

I see the problem now. I'll address it.

-Edward

On Thu, Mar 25, 2021 at 3:09 PM Gang ***@***.***> wrote:

> *@GangZhao-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/MISCLN.f
> <https://github.com/NOAA-EMC/EMC_post/pull/280#discussion_r601770207>:
>
> > +                 jj = jsta+j-1
> +                 do i=1,im
> +                   datapd(i,j,cfld) = GRID1(i,jj)
> +                 enddo
> +               enddo
> +             endif
> +           ENDIF
> +
> +!Hail parameter
> +            IF (IGET(993)>0) THEN
> +            DO J=JSTA,JEND
> +               DO I=1,IM
> +!               LAPSE=-((T700(I,J)-T500(I,J))/(D50*(Z700(I,J)-Z500(I,J))))
> +                LAPSE=(T700(I,J)-T500(I,J)) - 7.
> +                if (LAPSE <= 0.)LAPSE=0.
> +                SHIP=(MUCAPE(I,J)*D1000*Q1D(I,J)*LAPSE*(T500(I,J))*FSHR(I,J))/HCONST
>
> Hi Ed,
> The Q1D I mentioned in last comment is this Q1D used to compute SHIP. In
> last comment, I put the comment to the Q1S used to compute mixed layer (or
> LFC?)_, sorry for the confusion.
> So in this code block to compute SHIP, where does this Q1D come from?
> Since no CALCAPE or CALCAPE2 is called in this "IF (IGET(993)>0) THEN"
> block, so it has to come from somewhere before this IF-block, like just
> from the IF-block above for IGET(992) for mixed layer (or LFC?).
> Thanks!
>
>    - Gang
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/280#pullrequestreview-621488481>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQEG7SXW7BGGMVFU7XDTFOC6RANCNFSM4ZHVHYSQ>
> .
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Update MISCLN.f

Revised the Q1D variable used in the calculation of the hail parameter as well as deallocated several arrays that would have caused memory leaks at run time.

EdwardColon-NOAA committed:
Update MISCLN.f

AnnetteGibbs-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3736,6 +3758,323 @@ SUBROUTINE MISCLN
+
+            DO J=JSTA,JEND
+               DO I=1,IM
+                IF (MLLCL(I,J)>D2000) THEN
+                        MLLCLtmp=D00
+                ELSEIF (MLLCL(I,J)<=D1000) THEN
+                        MLLCLtmp=1.0
+                ELSE
+                        MLLCLtmp=((D2000-MLLCL(I,J))/D1000)
+                ENDIF
```
Hi Ed, do we have a surface based LCL we can use here instead of mixed-layer LCL for the Fixed Layer Tornado Parameter?  Just commenting so I don't prevent the merge.  We can discuss and look into this later.

EdwardColon-NOAA committed:
Update MISCLN.f

EdwardColon-NOAA committed:
Update MISCLN.f

EdwardColon-NOAA committed:
Update postxconfig-NT-3drtma.txt

WenMeng-NOAA commented on sorc/ncep_post.fd/CALHEL3.f:
```diff
@@ -0,0 +1,480 @@
+!> @file
```
@EdwardColon-NOAA This new routine has similarity with CALHEL2.f. Will you consider to unify them in the future?

EdwardColon-NOAA commented:
Yes. I developed a separate subroutine to avoid the complication but a
unified routine that computes both the fixed level and effective layer
helicity depending on the need would be the most practical choice going
forward.

-Edward

On Mon, Mar 29, 2021 at 8:03 PM WenMeng-NOAA ***@***.***>
wrote:

> *@WenMeng-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/CALHEL3.f
> <https://github.com/NOAA-EMC/EMC_post/pull/280#discussion_r603690410>:
>
> > @@ -0,0 +1,480 @@
> +!> @file
>
> @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA> This new routine
> has similarity with CALHEL2.f. Will you consider to unify them in the
> future?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/280#pullrequestreview-623741495>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQHG4OV63RU2KZU3UDLTGEINHANCNFSM4ZHVHYSQ>
> .
>


-- 
-------------------------------------------------
Edward Colón
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -282,9 +297,15 @@ SUBROUTINE MISCLN
-
-         IF(IGET(430) > 0) THEN
+! 0-6 km shear magnitude
 !$omp parallel do private(i,j,jj)
```
@EdwardColon-NOAA Please remove "jj" in line 301 which is not used in openmp code.

EdwardColon-NOAA committed:
Update MISCLN.f

WenMeng-NOAA commented:
The UPP standard-alone tests were conducted on WCOSS-Dell, WCOSS-Cray and Hera. No changed results were found. In the future, unifying routine  CALHEL2.f and CALHEL3.f could improve the code structure.
